### PR TITLE
PTP Dual nic support implementation.

### PR DESCRIPTION
**Pull request by aneeshkp**
PTP Dual nic support implmentation.
This PR adds support to handle dual nic support for boundary clock.
One of the nic will be running PTP without phc2sys.
The event framework handles various events based on logs from both nic ptp4l
and single phc2sys.

### CODE REVIEW

File: plugins/ptp_operator/config/config.go [hunk 1]
```diff
@@ -68,6 +69,36 @@ type PtpClockThreshold struct {
...
 
+// PtpProcessOpts ...
+type PtpProcessOpts struct {
+	// ptp4lOpts are set
+	Ptp4lOpts *string `json:"Ptp4lOpts,omitempty"`
+	// Phc2Opts options are set
+	Phc2Opts *string `json:"Phc2Opts,omitempty"`
+}
+
+// Ptp4lEnabled check if phc2sys is enabled
```
Comment (jzding) [inline plugins/ptp_operator/config/config.go:80]: nit: check if ptp4l is enabled

File: plugins/ptp_operator/metrics/manager.go [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+	lastClockState := stats.LastSyncState()
+	threshold := p.PtpThreshold(ptpProfileName)
+	switch clockState {
+	case ptp.LOCKED:
+		switch lastClockState {
+		case ptp.FREERUN: //last state was already sent for FreeRUN , but if its within then send again with new state
+			if isOffsetInRange(ptpOffset, threshold.MaxOffsetThreshold, threshold.MinOffsetThreshold) { // within range
+				log.Infof(" publishing event for ( profile %s) %s with last state %s and current clock state %s and offset %d for ( Max/Min Threshold %d/%d )",
+					ptpProfileName, eventResourceName, stats.LastSyncState(), clockState, ptpOffset, threshold.MaxOffsetThreshold, threshold.MinOffsetThreshold)
```
Comment (jzding) [inline plugins/ptp_operator/metrics/manager.go:216]: stats.LastSyncState() is called again here. Is it possible it returns a different value than L208?
Better use `lastClockState` from L208 instead.

File: plugins/ptp_operator/metrics/manager.go [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+				stats.SetLastOffset(int64(ptpOffset))
+				stats.AddValue(int64(ptpOffset)) // update off set when its in locked state and hold over only
+			}
+		case ptp.LOCKED: // last state was in sync , check if it is out of sync now
+			if isOffsetInRange(ptpOffset, threshold.MaxOffsetThreshold, threshold.MinOffsetThreshold) {
+				stats.SetLastOffset(int64(ptpOffset))
+				stats.AddValue(int64(ptpOffset)) // update off set when its in locked state and hold over only
+			} else {
+				log.Infof(" publishing event for ( profile %s) %s with last state %s and current clock state %s and offset %d for ( Max/Min Threshold %d/%d )",
+					ptpProfileName, eventResourceName, stats.LastSyncState(), clockState, ptpOffset, threshold.MaxOffsetThreshold, threshold.MinOffsetThreshold)
```
Comment (jzding) [inline plugins/ptp_operator/metrics/manager.go:228]: same comment here:  use `lastClockState` from L208.

File: plugins/ptp_operator/metrics/logparser.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+	if len(match) == 2 {
+		return match[0]
+	}
+	return -1
+}
+
+func isOffsetInRange(ptpOffset, maxOffsetThreshold, minOffsetThreshold int64) bool {
+	if ptpOffset < maxOffsetThreshold && ptpOffset > minOffsetThreshold {
+		return true
+	} else if ptpOffset > maxOffsetThreshold || ptpOffset < minOffsetThreshold {
```
Comment (jzding) [inline plugins/ptp_operator/metrics/logparser.go:234]: no need for `else if` here.

File: plugins/ptp_operator/config/config.go (hunk 1 above)
Comment (aneeshkp) [reply plugins/ptp_operator/config/config.go:80]: fixed

File: plugins/ptp_operator/metrics/manager.go (hunk 2 above)
Comment (aneeshkp) [reply plugins/ptp_operator/metrics/manager.go:216]: fixed

File: plugins/ptp_operator/metrics/manager.go [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+	switch clockState {
+	case ptp.LOCKED:
+		switch lastClockState {
+		case ptp.FREERUN: //last state was already sent for FreeRUN , but if its within then send again with new state
+			if isOffsetInRange(ptpOffset, threshold.MaxOffsetThreshold, threshold.MinOffsetThreshold) { // within range
+				log.Infof(" publishing event for ( profile %s) %s with last state %s and current clock state %s and offset %d for ( Max/Min Threshold %d/%d )",
+					ptpProfileName, eventResourceName, stats.LastSyncState(), clockState, ptpOffset, threshold.MaxOffsetThreshold, threshold.MinOffsetThreshold)
+				p.PublishEvent(clockState, ptpOffset, eventResourceName, eventType) // change to locked
+				stats.SetLastSyncState(clockState)
+				stats.SetLastOffset(int64(ptpOffset))
```
[the remark this answers is not in the document]
Comment (aneeshkp) [reply plugins/ptp_operator/metrics/manager.go]: fixed, we need to introduce go linter or revive to catch all these.

File: plugins/ptp_operator/metrics/manager.go (hunk 3 above)
Comment (aneeshkp) [reply plugins/ptp_operator/metrics/manager.go:228]: fixed

File: plugins/ptp_operator/metrics/logparser.go (hunk 4 above)
Comment (aneeshkp) [reply plugins/ptp_operator/metrics/logparser.go]: fixed

File: plugins/ptp_operator/metrics/manager.go (hunk 3 above)
Comment (jzding) [reply plugins/ptp_operator/metrics/manager.go:228]: I don't see it fixed

**Review by jzding — APPROVED** [review]

**Pull request merged**